File: src/ros2_bridge/mappers/nav_msgs/odometry.rs
//! Typed mapper for `nav_msgs/msg/Odometry`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn odometry_to_bus(msg: ros_env::nav_msgs::msg::Odometry) -> crate::nav_msgs::msg::v1::Odometry {
    crate::nav_msgs::msg::v1::Odometry {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        child_frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.child_frame_id),
        pose: Some(crate::ros2_bridge::mappers::geometry_msgs::pose_with_covariance::pose_with_covariance_to_bus(msg.pose)),
        twist: Some(crate::ros2_bridge::mappers::geometry_msgs::twist_with_covariance::twist_with_covariance_to_bus(msg.twist)),
    }
}

pub(crate) fn odometry_to_ros(bus: crate::nav_msgs::msg::v1::Odometry) -> ros_env::nav_msgs::msg::Odometry {
    ros_env::nav_msgs::msg::Odometry {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        child_frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.child_frame_id),
        pose: crate::ros2_bridge::mappers::geometry_msgs::pose_with_covariance::pose_with_covariance_to_ros(bus.pose.unwrap_or_default()),
        twist: crate::ros2_bridge::mappers::geometry_msgs::twist_with_covariance::twist_with_covariance_to_ros(bus.twist.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct NavMsgsOdometryMapper;

impl TypedTopicMapper for NavMsgsOdometryMapper {
    type Ros = ros_env::nav_msgs::msg::Odometry;
    type Bus = crate::nav_msgs::msg::v1::Odometry;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(odometry_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(odometry_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/nav_msgs/odometry.py
"""Generated mapper for `nav_msgs/msg/Odometry`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.pose_with_covariance import pose_with_covariance_to_bus, pose_with_covariance_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.twist_with_covariance import twist_with_covariance_to_bus, twist_with_covariance_to_ros

def odometry_to_bus(msg):
    from robot_bus.nav_msgs.msg.v1 import Odometry as BusMsg

    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.child_frame_id = str(msg.child_frame_id)
    bus.pose.CopyFrom(pose_with_covariance_to_bus(msg.pose))
    bus.twist.CopyFrom(twist_with_covariance_to_bus(msg.twist))
    return bus


def odometry_to_ros(bus):
    from nav_msgs.msg import Odometry as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.child_frame_id = str(bus.child_frame_id)
    out.pose = pose_with_covariance_to_ros(bus.pose)
    out.twist = twist_with_covariance_to_ros(bus.twist)
    return out


class NavMsgsOdometryMapper:
    def type_name(self) -> str:
        return "nav_msgs/msg/Odometry"

    def ros_msg_type(self):
        from nav_msgs.msg import Odometry as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return odometry_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.nav_msgs.msg.v1 import Odometry as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return odometry_to_ros(bus)
